dnstable/query.c

```diff
@@ -426,6 +426,108 @@ query_iter_next(void *clos, struct dnstable_entry **ent)
 }
 
 static dnstable_res
+query_iter_next_ip(void *clos, struct dnstable_entry **ent)
+{
+	struct query_iter *it = (struct query_iter *) clos;
+	struct timespec expiry = {0,0}, now;
+
+	if (it->query->do_timeout) {
+		my_gettime(CLOCK_MONOTONIC_RAW, &expiry);
+		my_timespec_add(&(it->query->timeout), &expiry);
+	}
+
+	for (;;) {
+		bool pass = false;
+		dnstable_res res;
+		const uint8_t *key, *val;
+		size_t len_key, len_val;
+
+		if (it->query->do_timeout) {
+			my_gettime(CLOCK_MONOTONIC_RAW, &now);
+			if (my_timespec_cmp(&now, &expiry) >= 0)
+				return (dnstable_res_timeout);
+		}
+
+		if (mtbl_iter_next(it->m_iter, &key, &len_key, &val, &len_val) != mtbl_res_success)
+			return (dnstable_res_failure);
+		*ent = dnstable_entry_decode(key, len_key, val, len_val);
+		assert(*ent != NULL);
+
+		/* it->key2 != NULL implies that this is used only for prefix
+		 * and range lookups. */
+		if (it->query->do_rrtype && it->key2 != NULL) {
+			uint16_t rrtype;
+			res = dnstable_entry_get_rrtype(*ent, &rrtype);
+			if (res != dnstable_res_success) {
+				dnstable_entry_destroy(ent);
+				return (res);
+			}
+
+			if (rrtype != it->query->rrtype) {
+				assert(ubuf_size(it->key) <= len_key);
+
+				/* We make a new start point with the prefix
+				 * of the current key, plus the target rrtype.
+				 */
+				ubuf *new_key = ubuf_init(64);
+				size_t rrtype_len = mtbl_varint_length(it->query->rrtype);
+				size_t key_prefix_len = ubuf_size(it->key) - rrtype_len;
+				ubuf_append(new_key, key, key_prefix_len);
+				add_rrtype_to_key(new_key, it->query->rrtype);
+
+				/* Advance to the next key if there are no
+				 * more possible matches with this prefix. */
+				if (memcmp(ubuf_data(new_key), key, ubuf_size(new_key)) <= 0) {
+					/* This loop advances us to the next
+					 * IP address.  It handles the case
+					 * of 10.0.255.255 -> 10.1.0.0. */
+					for (uint8_t *ptr = ubuf_data(new_key) + key_prefix_len - 1;;) {
+						(*ptr)++;
+						if (*ptr) {
+							break;
+						/* Check for overflow. If so
+						 * there are no more keys. */
+						} else if (ptr == ubuf_data(new_key)) {
```

Comment:
Should this conditional be `else if (ptr == ubuf_data(new_key) + 1)` ?

It seems like this block will never run, because if we loop all the way to the beginning of the key we'll end up incrementing the type byte from `\x02` to `\x03`, which would put us into the middle of the `ENTRY_TYPE_RDATA_NAME_REV` keys, unless I'm mistaken.

Testing this is a bit hard, since there's a bug in the range/prefix calculation -- `ip4_incr()` and `ip6_incr()` aren't checked for overflow.